src/onewire.rs

```diff
@@ -0,0 +1,135 @@
+}
+
+impl<'a> Device<'a> {
+    fn new(device_handle: onewire_device_t) -> Self {
+        Self {
+            _device: device_handle,
+            _p: PhantomData,
+        }
+    }
+
+    /// get the handle of the bus the device is attached to
+    pub fn bus(&self) -> onewire_bus_handle_t {
+        self._device.bus
+    }
+
+    /// get the device address
+    pub fn address(&self) -> u64 {
+        self._device.address
+    }
+}
+
+/// wrapper around the device iterator to search for available devices on the bus
+pub struct DeviceSearch<'b> {
+    _search: onewire_device_iter_handle_t,
+    _p: PhantomData<&'b ()>, // holds the lifetime since the search is linked to the lifetime of the bus
+}
+
+impl<'b> DeviceSearch<'b> {
+    fn new(bus: &mut OneWireBusDriver) -> Result<Self, EspError> {
+        let mut my_iter: onewire_device_iter_handle_t = ptr::null_mut();
+
+        esp!(unsafe { onewire_new_device_iter(bus._bus, &mut my_iter) })?;
+
+        Ok(Self {
+            _search: my_iter,
+            _p: PhantomData,
+        })
+    }
+
+    /// Search for the next device on the bus and yield it.
+    fn next_device(&mut self) -> Result<Device<'b>, EspError> {
+        let mut next_onewire_device = onewire_device_t::default();
+        esp!(unsafe { onewire_device_iter_get_next(self._search, &mut next_onewire_device) })?;
+
+        Ok(Device::new(next_onewire_device))
+    }
+}
+
+impl<'b> Iterator for DeviceSearch<'b> {
+    type Item = Device<'b>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Ok(dev) = self.next_device() {
+            Some(dev)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'b> Drop for DeviceSearch<'b> {
+    fn drop(&mut self) {
+        esp!(unsafe { onewire_del_device_iter(self._search) }).unwrap();
+    }
+}
+
+pub struct OneWireBusDriver<'d> {
+    _bus: onewire_bus_handle_t,
+    _p: PhantomData<&'d ()>,
+}
+
+impl<'d> OneWireBusDriver<'d> {
+    /// Create a new One Wire driver on the allocated pin.
+    ///
+    /// The pin will be used as an open drain output.
+    pub fn new(
+        pin: impl Peripheral<P = impl crate::gpio::InputPin + crate::gpio::OutputPin> + 'd,
+    ) -> Result<Self, EspError> {
+        let mut bus: onewire_bus_handle_t = ptr::null_mut();
+
+        let pin = pin.into_ref().pin();
+        let bus_config = esp_idf_sys::onewire_bus_config_t { bus_gpio_num: pin };
+
+        let rmt_config = esp_idf_sys::onewire_bus_rmt_config_t { max_rx_bytes: 10 };
+
+        esp!(unsafe { onewire_new_bus_rmt(&bus_config, &rmt_config, &mut bus as _) })?;
+
+        Ok(Self {
+            _bus: bus,
+            _p: PhantomData,
+        })
+    }
+    /// send reset pulse to the bus
+    pub fn reset_bus(&mut self) -> Result<(), EspError> {
+        esp!(unsafe { onewire_bus_reset(self._bus) })
+    }
+
+    pub fn search(&mut self) -> Result<DeviceSearch, EspError> {
```

Comment:
This is also problematic. In fact, I don't see how you can have an immutable borrow of the bus driver in the device driver, and still maintain the "iterator" approach for the `search` method. If you make this method `&self` rather than `&mut self`, it might work on the surface, _but_:
* I'm not sure the underlying C bus driver supports multiple concurrent device searches. Does it?
* You'll have a bit of a "device driver identity" problem here in that two subsequent searches will return two copies of the same device driver for the same device on the bus.

In fact, I'm not sure anymore that the device driver (what you currently call `Device`) is really a driver. I mean, it superficially resembles the SPI interface, in that in SPI you also have a bus + devices on it, however in the SPI case the devices are in no way dynamic - each device has a CS (chip select) pin, so it cannot just come and go. And in a way, with SPI - rather than "searching" for devices - you pre-create as many device drivers as you "know" up-front devices you have - each with its own CS pin.

Perhaps, whether the "device" is really a driver or just an "information" that a device might currently be available on the bus depends on how you send/receive data to the device.
Which goes up to my previous comment that I don't actually see in your code any methods to send/receive data...?

Reply:
I'm not married to the immutable borrow of the `bus` driver in the `device`. I am actually hoping to capture the relationship just using lifetimes.

I don't think concurrent device searches make sense, even if they are supported. Here is the doc-string
```
 Create an iterator to enumerate the 1-Wire devices on the bus\n\n
 @param[in] bus 1-Wire bus handle\n
 @param[out] ret_iter Returned created device iterator\n 
 @return\n      
- ESP_OK: Create device iterator successfully\n      
- ESP_ERR_INVALID_ARG: Invalid argument\n     
- ESP_ERR_NO_MEM: No memory to create device iterator\n     
- ESP_FAIL: Other errors"
```
, which doesn't mention anything about con-current searches. I think we should dis-allow them.

The "device" implementation is just information, you are correct. Every `OneWire` device has a unique identifier. Within that identifier is a "Family key" that allows the `OneWire` bus transmitter to identify the kind of device it is. When the C-API "finds" a device on the bus, it saves it's unique identifier into the `onewire_device_t` struct and returns it from the device search. 

That said, I don't think it is an issue for two subsequent searches to "find" the same device. While it's not typical, it is certainly possible for this to be hot-pluggable, and one might connect several devices to the bus, doing a search each time. Each of those search results is a valid device.

Reply:
> I'm not married to the immutable borrow of the `bus` driver in the `device`. I am actually hoping to capture the relationship just using lifetimes.

I think some of the lifetime issues in your sample gist code come from this ^^^ statement. The purpose of the `Borrow` trait and the `T` generic is to avoid using direct `&` references to the bus driver in the first place. This way the user pays the price of dealing with the `T` generic when using the driver, but she is able to:
- Pass a regular `&bus_driver` reference at the place of `T`
- Pass a smart pointer reference (`Rc` or `Arc`) at the place of `T` (depending on whether the driver is `Sync` or not)
- Pass the whole driver by moving (i.e. without a reference)

This is what `Borrow` gives you. But then again, the whole point is moot, because we probably don't need `Device` to become a full-scale driver, but rather - just an informational struct without IO capabilities and `'static` interior.

> I don't think concurrent device searches make sense, even if they are supported. Here is the doc-string

Fortunately a moot point once `Device` becomes a simple data carrier object.

> The "device" implementation is just information, you are correct. Every `OneWire` device has a unique identifier. Within that identifier is a "Family key" that allows the `OneWire` bus transmitter to identify the kind of device it is. When the C-API "finds" a device on the bus, it saves it's unique identifier into the `onewire_device_t` struct and returns it from the device search.
> 
> That said, I don't think it is an issue for two subsequent searches to "find" the same device. While it's not typical, it is certainly possible for this to be hot-pluggable, and one might connect several devices to the bus, doing a search each time. Each of those search results is a valid device.

Yeah, we are going the "information" way it seems...

Reply:
> This is what Borrow gives you. But then again, the whole point is moot, because we probably don't need Device to become a full-scale driver, but rather - just an informational struct without IO capabilities and 'static interior.

If a device does not have IO capabilities, I imagine a "device specific driver" will need to wrap the bus without owning it, so that multiple devices can mutably use the bus. The device specific driver will also need to wrap the "simple data carrier `Device`" so that it knows which device it is addressing.

Reply:
Why? Why can't the user just use the bus driver for sending / receiving data to a device? Why do we need a "device driver"?
- we do not have it in i2c which is the closest example
- we DO have it in spi, but there it is a necessity, as each device on the spi bus "owns" a cs pin (a peripheral)

Reply:
> Why can't the user just use the bus driver for sending / receiving data to a device?

They can, and it's completely fine for someone to do raw communication in this way. However, any device on the bus will have unique functions, error codes, registers and the like. The raw communication is not that useful on it's own, but this is more of a user concern than an `esp-idf-hal` concern.

Reply:
How you'll model the "upper layer" drivers is completely up to you, and yes, you probably need at least a struct-per-device-type. But this _does not mean_ these would need each their own OW "device" driver. They would just use a shared bus driver.

Reply:
@DaneSlattery I was re-reading the conversation and let me summarize, just to make sure we are on the same page:

> If a device does not have IO capabilities, I imagine a "device specific driver" will need to wrap the bus without owning it, so that multiple devices can mutably use the bus. 

Yes, this is a common problem. The `read`/`write`/`reset` methods on the bus driver will be `&mut` of course.
(
BTW, maybe we should just name it `OWDriver` as it is clear it would be a "bus" driver because there will be _no_ other OW driver in `esp-idf-hal` anyway. For I2C we did the same - `I2CDriver` is a "bus" driver, yet we just name it `I2CDriver`.
)

As for the common problem - we don't offer any native solution to this - it is up to the higher-level device-specific drivers to share the bus somehow - including when using the `I2C` driver. The easiest and natural way to do this of course is to put the `OWDriver` behind a `Mutex`, and e.g. the ["shared bus"](https://docs.rs/shared-bus/latest/shared_bus/) solution for I2C does exactly that.

We can implement something like `OWSharedBusDriver` (might even be a trait) that does it for the user, yet I feel we first need to nail down the main (non-shared) read/write/reset API.

> The device specific driver will also need to wrap the "simple data carrier Device" so that it knows which device it is addressing. I will work on an implementation of a device specific driver as an example.

This is where I'm a bit unclear what you mean. The device specific "driver" just needs some sort of **address**. And if I look at the API of the current `Device` structure you expose, it has just two items:
- A reference to the bus driver. This we absolutely don't need to be returned in the `Device` struct because it is _absolutely clear_ what bus we are talking about - this is the bus that returned the `Device` info after all, via its `search` method
- A `u64` address. This one we need, and this is maybe the only thing a device specific driver would need in addition to a (mutable reference to) the bus driver. So I would suggest:
  - A device specific "driver" takes on constructor (_usually_ - up to the device specific "Driver" implementor)
    - A `u64` that designates the device address
    - A mutable reference (or rather and best - a `T: BorrowMut<OWDriver<'a>>`) of the bus driver
  - A device specific "driver" does NOT therefore even need to know about the `Device` struct. It will be just an informational data carrier. See below.

Additionally:
- Let's turn the `Device` struct into a simple Rust newtype around `u64` (the address). Or even the `DeviceSearch` iterator could just return `u64`s - the addresses of the devices on the bus. 
- We don't really need the "bus driver" info which is currently in the `Device` struct. It is absolutely clear, on what bus this device is, simply because you call `search` on a **specific** bus driver

So something like:
```rust
impl<'a> OWDriver<'a> {
    pub fn new(...) {}

    pub fn search(&mut self) -> Result<impl Iterator<Item = u64>, EspError> {}

    pub fn reset(&mut self) -> Result<(), EspError> {}

    // Passing `None` for an address means broadcast the write to all devices connected to the bus
    pub fn write(&mut self, addr: Option<&u64>, data: &[u8]) -> Result<(), EspError> {}

    // Here I'm a bit unsure as I haven't checked in details how read actually works, and if this signature is good enough
    pub fn read(&mut self, addr: &u64, buf: &mut [u8]) -> Result<usize, EspError> {}
}
```

?

Reply:
I have just pushed an update implementing your changes. Let me know how you feel about the design I went with.

Reply:
Yes - I think the design is good. Once you address the various nits (and remove `OWDevice` or describe why it is still necessary) I think we should be in a good shape for merging. Thanks a lot, and markind this as resolved.